storage/cleanup.go

```diff
@@ -132,38 +131,47 @@ func (m *cleanupManager) cleanupFilesetFiles(earliestToRetain time.Time) error {
 // writes and future writes, we need to shift flush time range by block size as the
 // time range for commit log files we need to check.
 func (m *cleanupManager) commitLogTimeRange(t time.Time) (time.Time, time.Time) {
-	flushStart, flushEnd := m.fm.FlushTimeStart(t), m.fm.FlushTimeEnd(t)
-	return flushStart.Add(-m.blockSize), flushEnd.Add(-m.blockSize)
+	var (
+		ropts      = m.opts.CommitLogOptions().RetentionOptions()
+		blockSize  = ropts.BlockSize()
+		flushStart = retention.FlushTimeStart(ropts, t)
+		flushEnd   = retention.FlushTimeEnd(ropts, t)
+	)
+	return flushStart.Add(-blockSize), flushEnd.Add(-blockSize)
 }
 
 // commitLogTimes returns the earliest time before which the commit logs are expired,
 // as well as a list of times we need to clean up commit log files for.
 func (m *cleanupManager) commitLogTimes(t time.Time) (time.Time, []time.Time) {
-	earliest, latest := m.commitLogTimeRange(t)
-
-	// TODO(xichen): preallocate the slice here
-	var commitLogTimes []time.Time
-	for commitLogTime := latest; !commitLogTime.Before(earliest); commitLogTime = commitLogTime.Add(-m.blockSize) {
-		hasFlushedAll := true
-		leftBlockStart := commitLogTime.Add(-m.blockSize)
-		rightBlockStart := commitLogTime.Add(m.blockSize)
-		for blockStart := leftBlockStart; !blockStart.After(rightBlockStart); blockStart = blockStart.Add(m.blockSize) {
-			if m.fm.NeedsFlush(blockStart) {
-				hasFlushedAll = false
-				break
-			}
-		}
-		if hasFlushedAll {
-			commitLogTimes = append(commitLogTimes, commitLogTime)
-		}
-	}
-
-	return earliest, commitLogTimes
+	var (
+		ropts            = m.opts.CommitLogOptions().RetentionOptions()
+		blockSize        = ropts.BlockSize()
+		earliest, latest = m.commitLogTimeRange(t)
+	)
+	// NB(prateek): this logic of polling the namespaces across the commit log's entire
+	// retention history could get expensive if commit logs are retained for long periods.
+	// e.g. if we retain them for 40 days, with a block 2 hours; then every time
+	// we try to flush we are going to be polling each namespace, for each shard, for 480
+	// distinct blockstarts. Say we have 2 namespaces, each with 8192 shards, that's ~10M map lookups.
+	// If we cared about 100% correctness, we would optimize this by retaining a smarter data
+	// structure (e.g. interval tree), but for our use-case, it's safe to assume that commit logs
+	// are only retained for a period of 1-2 days (at most), after we which we'd live we with the
+	// data loss.
```

Comment:
It seems that this is what we are doing though no?  We check every possible block and determine if it needs a flush or not?

TBH a far more optimal approach would be to find all commit log files on disk, then specifically check for each of them whether there are block files covering those time ranges.

This approach of calculating all possible permutations to cleanup then check if commit logs exist at any of those points is definitely quite expensive.

Reply:
I think the code addresses that concern ~ this function returns (1) the earliest commit log time to retain, and (2) a list of commit log times after (1) which can be cleaned up because we've flushed the fileset files for them. `cleanupCommitLogs` (below) deletes all commit log files older than (1).